# 👥 Two Sample T-Test in Statistics
### Comparing Two Independent Groups  

---

> **Two Sample T-Test** is used to determine whether the **means of two independent groups** are significantly different from each other.  
> Unlike the one-sample T-Test, here we are **not comparing to a population mean** — we are comparing **two separate groups** against each other.

---

## 🤔 One Sample vs Two Sample T-Test

| | One Sample T-Test | Two Sample T-Test |
|---|---|---|
| Comparing | Sample mean vs Population mean | Group A mean vs Group B mean |
| Groups | 1 | 2 |
| Use Case | "Is our sample different from the known population?" | "Are these two groups different from each other?" |

---

## 📌 Problem Statement

A teacher wants to check if **Group A students** perform significantly better than **Group B students** in a math test.

| Detail | Value |
|---|---|
| Group A | 10 students |
| Group B | 10 students |
| Significance Level (α) | 0.05 |
| equal_var | False (Welch's T-Test) |

---

## 🔢 Formula

$$t = \frac{\bar{x}_A - \bar{x}_B}{\sqrt{\frac{s_A^2}{n_A} + \frac{s_B^2}{n_B}}}$$

- $\bar{x}_A$, $\bar{x}_B$ = Mean of Group A and Group B  
- $s_A$, $s_B$ = Standard Deviation of Group A and Group B  
- $n_A$, $n_B$ = Sample sizes  

> **equal_var=False** means we use **Welch's T-Test** — which does NOT assume both groups have equal variance. This is safer and more realistic for real-world data.

In [1]:
# Importing required libraries
import numpy as np
from scipy import stats

## 📋 Step 1 — Define the Two Groups

We define the math scores of both groups and explore their basic statistics.

In [3]:
group_A = [85, 88, 90, 92, 87, 85, 89, 91, 86, 88]
group_B = [82, 84, 80, 83, 81, 79, 78, 85, 84, 83]

print("-" * 40)
print(f"Group A Scores  : {group_A}")
print(f"Group A Mean    : {np.mean(group_A):.2f}")
print(f"Group A Std Dev : {np.std(group_A, ddof=1):.2f}")
print("-" * 40)
print(f"Group B Scores  : {group_B}")
print(f"Group B Mean    : {np.mean(group_B):.2f}")
print(f"Group B Std Dev : {np.std(group_B, ddof=1):.2f}")
print("-" * 40)
print(f"Mean Difference : {np.mean(group_A) - np.mean(group_B):.2f}")

----------------------------------------
Group A Scores  : [85, 88, 90, 92, 87, 85, 89, 91, 86, 88]
Group A Mean    : 88.10
Group A Std Dev : 2.42
----------------------------------------
Group B Scores  : [82, 84, 80, 83, 81, 79, 78, 85, 84, 83]
Group B Mean    : 81.90
Group B Std Dev : 2.33
----------------------------------------
Mean Difference : 6.20


## 🔍 Step 2 — Perform the Two Sample T-Test

We use `stats.ttest_ind()` from SciPy with `equal_var=False` to run **Welch's T-Test**.

> **Why equal_var=False?**  
> When we don't know if both groups have the same variance (which is usually the case in real life), Welch's T-Test is preferred over the standard Student's T-Test.

In [4]:
t_stats, p_value = stats.ttest_ind(group_A, group_B, equal_var=False)

print(f"T-Statistic : {t_stats:.4f}")
print(f"P-Value     : {p_value:.6f}")

T-Statistic : 5.8296
P-Value     : 0.000016


## ✅ Step 3 — Decision: Accept or Reject Null Hypothesis?

### Our Hypotheses:
- **H₀ (Null Hypothesis):** Mean of Group A = Mean of Group B *(no significant difference)*
- **H₁ (Alternative Hypothesis):** Mean of Group A ≠ Mean of Group B *(significant difference exists)*

### Decision Rule:

| Condition | Decision |
|---|---|
| p_value < α (0.05) | ❌ Reject H₀ |
| p_value ≥ α (0.05) | ✅ Accept H₀ |

In [5]:
alpha = 0.05

print(f"T-Statistic : {t_stats:.4f}")
print(f"P-Value     : {p_value:.6f}")
print(f"Alpha       : {alpha}")
print("-" * 40)

if p_value < alpha:
    print(" Reject the Null Hypothesis")
    print("→ There IS a significant difference between Group A and Group B.")
else:
    print(" Accept the Null Hypothesis")
    print("→ No significant difference found between Group A and Group B.")

T-Statistic : 5.8296
P-Value     : 0.000016
Alpha       : 0.05
----------------------------------------
 Reject the Null Hypothesis
→ There IS a significant difference between Group A and Group B.


## 📝 Conclusion

| Metric | Value |
|---|---|
| Group A Mean | 88.1 |
| Group B Mean | 81.9 |
| Mean Difference | 6.2 |
| T-Statistic | 5.8296 |
| P-Value | ~0.0001 |
| Alpha (α) | 0.05 |
| Decision | ❌ Reject H₀ |

---

**Interpretation:**  
Since **p-value (~0.0001) < alpha (0.05)**, we **reject the null hypothesis**.  
There is **strong statistical evidence** that Group A performs significantly better than Group B.

---

## 🔁 All Three Tests — Final Comparison

| | Z-Test | One Sample T-Test | Two Sample T-Test |
|---|---|---|---|
| Comparing | Sample vs Population | Sample vs Population | Group A vs Group B |
| Std Dev | Population Known | Sample (ddof=1) | Both Groups |
| Test Statistic | 1.3703 | 1.7782 | 5.8296 |
| P-Value | 0.1706 | ~0.1092 | ~0.0001 |
| Decision | ✅ Accept H₀ | ✅ Accept H₀ | ❌ Reject H₀ |

> 💡 **Key Takeaway:** A **very high T-Statistic (5.83)** and **very low P-Value (~0.0001)** means the two groups are clearly different — this is a strong rejection of H₀. The bigger the T-Score, the more confident we are that the difference is real and not due to random chance.

